internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -451,6 +451,9 @@ func (r *Reconciler) scaleMachineSet(ctx context.Context, ms *clusterv1.MachineS
 	// Mutate replicas and the related annotation.
 	ms.Spec.Replicas = &newScale
 	mdutil.SetReplicasAnnotations(ms, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+mdutil.MaxSurge(*deployment))
+	if _, err := mdutil.SetOldMachineDeploymentAnnotations(ms, deployment.Annotations[clusterv1.RevisionAnnotation]); err != nil {
```

Comment:
What's the rationale for this to. be here? Have you consider to have node taining to be part of the reconcileOldMachineSets logic?

Reply:
As some people already mentioned the label propagation, maybe there's no need to explain, but my first idea was to taint from machineDeploymentController. However, I came to realize that machineDeploymentController is not designed to operate with node directly. We need to introduce the watcher which will change the structure of machineDeploymentController and need to add some helper functions which are already defined in the other places (i.e., not DRY). That's why I decided to use label propagation.